### PR TITLE
Ros2 dashing

### PR DESCRIPTION
**Pull request by vandanamandlik**
@AAlon , Please find first PR for Deep racer application.

*Description of changes:*
**Deepracer simulation migration from ROS1 to ROS2**
- Modified CMakeLists.txt and package.xml as per ROS2 guidelines.
- Added ROS2 Deep Racer Simulation package dependencies repos to .rosinstall file.
- Added launch file to launch racer car and car track model in gazebo.
- converted .xacro file of car model to sdf file.

**Steps to build and test simulation workspace locally:**
- First add COLCON_IGNORE empty file at following paths to avoid build error as these packages are not part of this PR.
  - simulation_ws/src/deepracer_interpreter/COLCON_IGNORE
  - simulation_ws/src/deepracer_msgs/COLCON_IGNORE
  - simulation_ws/src/sagemaker_rl_agent/COLCON_IGNORE
- cd simulation_ws
- rosws update
- rosdep install --from-paths src --ignore-src -r -y
- colcon build
- export GAZEBO_MODEL_PATH=$GAZEBO_MODEL_PATH:<GAZEBO_MODELS_PATH_HERE>
   eg. export GAZEBO_MODEL_PATH=$GAZEBO_MODEL_PATH:/home/root-pc/vandana/ros2/deep_exm_ws/src/deepracer_simulation/models
- export LD_LIBRARY_PATH=$LD_LIBRARY_PATH:/usr/lib/x86_64-linux-gnu/gazebo-9/plugins/
- ros2 launch deepracer_simulation racecar_simulation.launch.py

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -1,18 +1,17 @@
...
     on_failure: always
     recipients:
     - ros-contributions@amazon.com
+    #- travis-build@platform-notifications.robomaker.aws.a2z.com Uncomment once the build passes
 env:
   matrix:
-  - ROS_DISTRO=kinetic ROS_VERSION=1 GAZEBO_VERSION=7
-  - ROS_DISTRO=kinetic ROS_VERSION=1 GAZEBO_VERSION=9
-  - ROS_DISTRO=melodic ROS_VERSION=1 GAZEBO_VERSION=9
+  - ROS_DISTRO=dashing ROS_VERSION=2 GAZEBO_VERSION=9 WORKSPACES="robot_ws"
```
Comment (AAlon) [inline .travis.yml]: Remove - deepracer doesn't have `robot_ws`

File: simulation_ws/src/deepracer_simulation/CMakeLists.txt [hunk 2]
```diff
@@ -1,19 +1,10 @@
...
-catkin_package()
-
-###########
-## Build ##
-###########
-
-include_directories(
-  ${catkin_INCLUDE_DIRS}
-)
+find_package(ament_cmake_ros REQUIRED)
```
Comment (AAlon) [inline simulation_ws/src/deepracer_simulation/CMakeLists.txt]: ament_cmake, not ament_cmake_ros

File: simulation_ws/src/deepracer_simulation/CMakeLists.txt [hunk 3]
```diff
@@ -24,24 +15,25 @@ install(PROGRAMS
...
+if(BUILD_TESTING)
+  find_package(ament_lint_auto REQUIRED)
+  set(ament_cmake_copyright_FOUND TRUE)
+  ament_lint_auto_find_test_dependencies()
+endif()
+
+################################################################################
+# Macro for ament package
+################################################################################
+ament_export_dependencies(rclpy)
```
Comment (AAlon) [inline simulation_ws/src/deepracer_simulation/CMakeLists.txt]: I don't think you need to export anything from deepracer_simulation?

File: simulation_ws/src/deepracer_simulation/package.xml [hunk 4]
```diff
@@ -8,26 +8,29 @@
...
-  <buildtool_depend>catkin</buildtool_depend>
+  <buildtool_depend>ament_cmake</buildtool_depend>
+  <depend>gazebo_ros</depend>
+  <depend>rclpy</depend>
+  <depend>gazebo_plugins</depend>
+  <depend>gazebo_ros_pkgs</depend>
+  <exec_depend>gazebo</exec_depend>
 
-  <build_depend>ackermann_msgs</build_depend>
+  <!-- TODO build_depend>ackermann_msgs</build_depend>
```
Comment (AAlon) [inline simulation_ws/src/deepracer_simulation/package.xml:18]: Why are these commented out?

File: simulation_ws/src/deepracer_simulation/CMakeLists.txt [hunk 5]
```diff
@@ -24,24 +15,25 @@ install(PROGRAMS
...
-)
-install(DIRECTORY models/
-  DESTINATION ${CATKIN_PACKAGE_SHARE_DESTINATION}/models
-)
-install(DIRECTORY urdf/
-  DESTINATION ${CATKIN_PACKAGE_SHARE_DESTINATION}/urdf
-)
-install(DIRECTORY worlds/
-  DESTINATION ${CATKIN_PACKAGE_SHARE_DESTINATION}/worlds
+install(DIRECTORY launch config models worlds
```
Comment (AAlon) [inline simulation_ws/src/deepracer_simulation/CMakeLists.txt:21]: How come you don't install meshes, urdf?

**Review by AAlon — CHANGES_REQUESTED** [review]
Thanks, got some minor comments. Could you also please add COLCON_IGNORE for packages that haven't been migrated yet? Let's get the CI to pass, so far it failed because of deepracer_msgs: https://travis-ci.org/aws-robotics/aws-robomaker-sample-application-deepracer/jobs/575357941

File: .travis.yml (hunk 1 above)
Comment (vandanamandlik) [reply .travis.yml]: OK removed.

File: simulation_ws/src/deepracer_simulation/CMakeLists.txt (hunk 2 above)
Comment (vandanamandlik) [reply simulation_ws/src/deepracer_simulation/CMakeLists.txt]: Done.

File: simulation_ws/src/deepracer_simulation/CMakeLists.txt (hunk 3 above)
Comment (vandanamandlik) [reply simulation_ws/src/deepracer_simulation/CMakeLists.txt]: Removed the exports.

File: simulation_ws/src/deepracer_simulation/package.xml (hunk 4 above)
Comment (vandanamandlik) [reply simulation_ws/src/deepracer_simulation/package.xml:18]: [WIP] Right now this functionality is in progress so commented.

File: simulation_ws/src/deepracer_simulation/CMakeLists.txt (hunk 5 above)
Comment (vandanamandlik) [reply simulation_ws/src/deepracer_simulation/CMakeLists.txt:21]: We have moved meshes to models folder. 
Converted urdf to sdf and moved it to models folder.

Comment (AAlon) [reply simulation_ws/src/deepracer_simulation/CMakeLists.txt:21]: Any reason for moving them? In general we'd like to change only the things that are needed for ROS2. If there is a reason (e.g. if in ROS2 the convention around directory structure has changed) that's fine, but right now it just seems like there are a bunch of changes that are not strictly needed. This makes reviewing more difficult.

Comment (vandanamandlik) [reply simulation_ws/src/deepracer_simulation/CMakeLists.txt:21]: Yes. They are moved as per ROS2 convention. So there are structure changes.
We took reference from following link:
https://github.com/ROBOTIS-GIT/turtlebot3_simulations/tree/ros2/turtlebot3_gazebo/models/turtlebot3_burger.

Updated structure is as follows:

/models
  - /easy_track
    - meshes
    	- scripts
    	- textures
    - model.config
    - model.sdf
  - /hard_track
    - dae/png/jpg files
    - model.config
    - model.sdf
  - /medium_track
    - meshes
    	- scripts
    	- textures
    - model.config
    - model.sdf
  - /racer_car
    - meshes
    - model.config
    - model.sdf

NOTE: Older launch files will be removed in the next PR.

File: simulation_ws/src/deepracer_simulation/models/hard_track/model.sdf [hunk 6]
```diff
@@ -5,14 +5,14 @@
       <collision name="collision">
         <geometry>
           <mesh>
-            <uri>model://deepracer_simulation/meshes/hard_track/Track3_and_Wall.dae</uri>
+            <uri>model://hard_track/Track3_and_Wall.dae</uri>
```
Comment (timrobotson) [inline simulation_ws/src/deepracer_simulation/models/hard_track/model.sdf]: Shouldn't these be in a meshes folder to be consistent?

File: simulation_ws/src/deepracer_simulation/CMakeLists.txt (hunk 5 above)
Comment (timrobotson) [reply simulation_ws/src/deepracer_simulation/CMakeLists.txt:21]: Why are the hard_track dae and image files not in meshes/textures directories like the rest?

File: simulation_ws/src/deepracer_simulation/models/hard_track/model.sdf (hunk 6 above)
Comment (vandanamandlik) [reply simulation_ws/src/deepracer_simulation/models/hard_track/model.sdf]: Yes. I will make that change.

File: simulation_ws/src/deepracer_simulation/CMakeLists.txt (hunk 5 above)
Comment (vandanamandlik) [reply simulation_ws/src/deepracer_simulation/CMakeLists.txt:21]: Actually in ROS1 code for hard_track, dae and images files were inside hard_track folder. There was not textures folder. I will make that change and commit it again.

File: simulation_ws/.rosinstall [hunk 7]
```diff
@@ -1 +1,3 @@
-
+# Amazon ROS2 Deep Racer Simulation package dependencies
+# Gazebo packages
+- git: {local-name: src/deps/gazebo/gazebo_ros_pkgs, uri: 'https://github.com/ros-simulation/gazebo_ros_pkgs.git', version: ros2}
```
Comment (AAlon) [inline simulation_ws/.rosinstall:3]: Remove gazebo_ros_pkgs as it's available via apt.

**Review by AAlon — APPROVED** [review]
Overall LGTM :) last comment from my side below. 

Full disclosure- we haven't sufficiently tested this yet so there might be more asks / issues we'll discover later.
 
Any estimate on the robot_ws part?

**Comment by vandanamandlik** [comment]
> Overall LGTM :) last comment from my side below.
> 
> Full disclosure- we haven't sufficiently tested this yet so there might be more asks / issues we'll discover later.
> OK. We will address it if you find any issues.
> Any estimate on the robot_ws part?
There is no robot_ws in deepracer ROS1 package.

File: simulation_ws/.rosinstall (hunk 7 above)
Comment (vandanamandlik) [reply simulation_ws/.rosinstall:3]: Some of the functionalities that we require here in deepracer are not available in released version of gazebo_ros_pkgs. So we have to add it to .rosinstall.

**Pull request merged**